### PR TITLE
Tidy up evaluation

### LINKED ISSUE #77: Serarate train and eval nicely
While doing #76 it became clear that the way we do eval is too intertwined with training.

This is a kinda exploratory issue but broadly the goal would be:

1. use only the eval dataloader for metrics and figures
2. eval on n batches, instead of the last microbatch of train data
3. remove existing usage of intermediate train variables from the train fwd pass evals
4. properly hook in `n_eval_steps` to be correctly used in metrics and figures (interpreted as the number of batches of the eval dataloader to use for metrics and figures)
5. allow setting eval_batch_size separately

### PR DESCRIPTION
**Pull request by oclivegriffin**
## Description
Features:
- separates train logging from running eval. we now log train metrics way more often than doing eval. new config options:
  - `train_log_freq`: How often to to log training metrics (loss, l0, n_alive)
  - `eval_freq`: How often to run evaluation (via "create_metrics")
    - `n_eval_steps`: The number of steps over which to run both eval and figure creation.
  - `eval_batch_size`
  - `print_freq` is replaced by both `eval_freq` and `train_log_freq`
- Multi-batch eval necessitates the introduction of a different way of structuring metrics and figures. They now need to be able to operate over multiple batches and accumulate. I've added  `StreamingMetricCreator` and `StreamingFigureCreator` for this. they `.watch()` multiple batches then `.compute()` a final set of metrics/figures
- rearranged how stuff is logged to be tidier
  - all plots now render by default with the convention row=module
-  Cleaner separation in wandb with train and eval logs.

## Related Issue
Closes #77 

## Motivation and Context
While doing https://github.com/goodfire-ai/spd/pull/76 it became clear that the way we do eval is too intertwined with training.

We want to:
- make it clear when we're using the eval / train set
- allow evaluating on not just the final microbatch of train data
- allow eval batch size to be different
- remove existing usage of intermediate train variables from the train fwd pass evals
- properly hook in n_eval_steps to be correctly used in metrics and figures

## How Has This Been Tested?
many manual runs

## Does this PR introduce a breaking change?
Yes, we no longer have figures_fns and metrics_fns in the config. These will need to be converted to `eval_metrics: list[EvalMetricConfig]`. There's another few config changes which give deprecation warnings.

### CODE REVIEW

**Comment by oli-clive-griffin** [comment]
test run: https://wandb.ai/goodfire/spd/reports/Test-eval-train-refactor--VmlldzoxMzcwNzYwMg==

**Comment by danbraunai** [comment]
@claude curious what you think of the changes made since your last comment? Ultrathink about the PR as a whole now. I want critical comments, do feel comfortable to give Oli the hard truths, he's the toughest guy I know!

File: spd/figures.py [hunk 1]
```diff
@@ -22,133 +23,158 @@
...
+            figure_creator.watch_batch(ci=ci, batch=batch)
+
+    out: dict[str, plt.Figure] = {}
+    all_dicts = [figure_creator.compute() for figure_creator in figure_creators]
+    for d in all_dicts:
+        if set(d.keys()).intersection(out.keys()):
+            raise ValueError(f"Keys {set(d.keys()).intersection(out.keys())} already in output")
+        out.update(d)
+
+    return out
```
Comment (danbraunai) [inline spd/figures.py:180]: Hmmm maybe we want to combine a lot of the figure and metric code? I think it makes sense in general for each of them to have the same inputs anyway (batch, target_out, ci).

Given that, StreamingMetricCreator and StreamingFigureCreator can be combined, as well as create_metrics and create_figures.

Fundamentally they're both metrics of a sort. The differences are that images take up more bytes, and typically take longer to produce, a human has a harder time viewing a lot of them. All of these lead to us wanting to run them less frequently, and save them differently, but not much else.

We could just have a single list of metrics which includes the figures, but a flag/config arg for "slow metrics" which would currently contain the figures, but might contain other things we don't care about running as frequently as the other metrics. Our config args could be `eval_freq` and `slow_eval_freq`, and maybe `slow_eval_on_first_step`.

Thoughts?

File: spd/run_spd.py [hunk 2]
```diff
@@ -28,6 +28,17 @@
 from spd.utils.run_utils import save_file
 
 
+def loop_dl[T](dl: DataLoader[T]):
```
Comment (danbraunai) [inline spd/run_spd.py:31]: I prefer `loop_dataloader`. It's short enough imo

**Review by danbraunai — CHANGES_REQUESTED** [review]
Much better. Though I made a big comment about combining metrics and figures which I think can make it better. Curious on your thoughts there.

Also, as mentioned above, would be good to run on 20 tms 5-2 seeds and see that it solves it 18/20 times (or around there) like [here](https://github.com/goodfire-ai/spd/pull/65#pullrequestreview-3035892036).

File: spd/configs.py [hunk 3]
```diff
@@ -22,62 +21,63 @@
...
-    extra_kwargs: dict[str, Any] = Field(
+    extra_init_kwargs: dict[str, Any] = Field(
         default={},
-        description="Extra keyword arguments to pass to the function besides the default `inputs`",
+        description="Extra keyword arguments to pass to the class constructor besides `model: ComponentModel` and `config: Config`",
     )
 
     @abstractmethod
-    def get_real_func(self) -> Callable[..., Any]: ...
+    def get_real_class(self) -> type[Any]: ...
```
[the remark this answers is not in the document]
Comment (oli-clive-griffin) [reply spd/configs.py]: changed to `_get_metric_class` in line with the wider changes

File: spd/figures.py (hunk 1 above)
Comment (oli-clive-griffin) [reply spd/figures.py]: for posterity:

We talked in person and agreed this was good. I've gone with a unified approach and a slow flag on slow metrics which are logged every `slow_eval_freq` steps.

**Comment by oli-clive-griffin** [comment]
@claude can you summarise the current state of this PR in a way that i can paste in the PR description. same style as my original PR description ideally

**Comment by oli-clive-griffin** [comment]
@claude we're finding that after these changes, our training is less stable, however these changes shouldn't actually impact training dynamics meaningfully, can you deeply look at this PR's implications for training and see if there's a change here that would explain impacted training, specifically with the autograd computational graph and that kind of thing

**Comment by danbraunai** [comment]
> did a large seed sweep here too: https://wandb.ai/goodfire/spd/reports/seed-sweep--VmlldzoxMzczNDI4NA

- tms_5-2 on this sweep fails 5/20 broken (see above) compared to 2 broken on dev ([[here](https://wandb.ai/goodfire/spd?nw=j1h3xn3t52b)](https://wandb.ai/goodfire/spd?nw=j1h3xn3t52b))
- tms_40-10 has 6/19 broken ([[here](https://wandb.ai/goodfire/spd/workspace?nw=yla65nc2na7&panelDisplayName=eval%2Ffigures%2Fcausal_importances_upper_leaky&panelSectionName=eval%2Ffigures)](https://wandb.ai/goodfire/spd/workspace?nw=yla65nc2na7&panelDisplayName=eval%2Ffigures%2Fcausal_importances_upper_leaky&panelSectionName=eval%2Ffigures)) compared to 4/19 broken on dev ([[here](https://wandb.ai/goodfire/spd/panel/7lnelfsjn?nw=jhx2vzndpn0)](https://wandb.ai/goodfire/spd/panel/7lnelfsjn?nw=jhx2vzndpn0))

Not ideal :(.

100 seeds experiment

- oli/eval-train tms_5-2 [[here](https://wandb.ai/goodfire/spd?nw=l9pvlmjxm05)](https://wandb.ai/goodfire/spd?nw=l9pvlmjxm05). (tried to run 100 but only 90 there). 4/30 + 8/30 + 6/30 = **18/90** broken
- dev tms_5-2 [[here](https://wandb.ai/goodfire/spd?nw=ab34oe32rbc)](https://wandb.ai/goodfire/spd?nw=ab34oe32rbc) (filtered to first 90 runs) 4/30 + 7/30 + 2/30 = **13/90** broken (+3/10 for the last 10)
- oli/eval-train tms_40-10 [[here](https://wandb.ai/goodfire/spd?nw=tgnl42t2tt9)](https://wandb.ai/goodfire/spd?nw=tgnl42t2tt9). (again only 90 runs) 13/30 + 5/30 + 6/30 = **24/90** broken
- dev tms_40-10 [[here](https://wandb.ai/goodfire/spd?nw=g1l7zvi6lg1)](https://wandb.ai/goodfire/spd?nw=g1l7zvi6lg1) (filtered to first 90 runs) 9/30 + 3/30 + 7/30 = **19/90** broken (+3/10 for the last 10)


fwiw I did some experiments with higher C, but that's tangential to the point of whether this PR breaks things:
- With C=30 on this PR we can get 2/20 broken ([[here](https://wandb.ai/goodfire/spd?nw=87zp4e2nasl&panelDisplayName=eval%2Ffigures%2Fcausal_importances_upper_leaky&panelSectionName=eval%2Ffigures)](https://wandb.ai/goodfire/spd?nw=87zp4e2nasl&panelDisplayName=eval%2Ffigures%2Fcausal_importances_upper_leaky&panelSectionName=eval%2Ffigures))
- With C=300 on the branch we get 2/19 broken.

So bigger C seems to make it more stable, but that's a side point.

**Comment by oli-clive-griffin** [comment]
@claude we're finding that after these changes, our training is less stable, however these changes shouldn't actually impact training dynamics meaningfully, can you deeply look at this PR's implications for training and see if there's a change here that would explain impacted training, specifically with the autograd computational graph and that kind of thing

**Comment by danbraunai** [comment]
Just noting that once [PR38](https://github.com/goodfire-ai/spd/pull/38) is in (which it should be before this PR goes in), we'll want to rearrange where the functions implemented in that are defined and called (fine to not put it all in the metrics file if it's too big).

**Comment by danbraunai** [comment]
Noting some figures from Oli's runs of 200 seeds.

- 200 seeds of tms_5-2 [here](https://wandb.ai/goodfire/spd/workspace/panel/xjxrjpiis?nw=nnmgszurquq). Has 48/200 broken. This is compared to 13/90 (which is 28.9/200) for the old dev runs that I made further up in this thread.
- 200 seeds of tms_40-10 also [here](https://wandb.ai/goodfire/spd/workspace/panel/xjxrjpiis?nw=nnmgszurquq). Has 37/200 broken. This is compared to 19/90 (which is 42.2/200) for the old dev runs that I made further up in this thread.

Quite inconclusive there. For some reason, unless I've screwed up the comparisons, tms_5-2 gets a lot worse, but tms_40-10 is similar.

I'll probably merge in dev then set off both again for 200 seeds each on tms_5-2.

**Comment by danbraunai** [comment]
- Ran 200 seeds on latest dev (14da9488), got 48/200 broken on [tms_5-2](https://wandb.ai/goodfire/spd?nw=jwvc0799bzx).
- Also ran 200 seeds on latest of this branch, got 51/200 (compared to 48 last time).

I'm now very confident that this PR doesn't make the runs worse. I'll do a final review

File: spd/models/component_model.py [hunk 4]
```diff
@@ -202,7 +202,7 @@ def _patch_modules(
                     C=C,
                     d_in=d_in,
                     d_out=d_out,
-                    bias=module.bias.data if module.bias is not None else None,  # pyright: ignore[reportUnnecessaryComparison]
+                    bias=module.bias if module.bias is not None else None,  # pyright: ignore[reportUnnecessaryComparison]
```
[the remark this answers is not in the document]
Comment (danbraunai) [reply spd/models/component_model.py:205]: I reverted this change. Too risky for people to accidentally train it and I don't see the downside with just storing the tensor but not the parameter.
